catalog/core/catalog-core-subscriptionstore/subscriptionstore-impl/src/main/java/org/codice/ddf/catalog/subscriptionstore/common/CachedSubscription.java

```diff
@@ -0,0 +1,195 @@
+    private static final Logger LOGGER = LoggerFactory.getLogger(CachedSubscription.class);
+
+    private static final String SUBSCRIPTION_ID_OSGI = "subscription-id";
+
+    private static final String EVENT_ENDPOINT = "event-endpoint";
+
+    private static final Integer INITIAL_CAPACITY = 4;
+
+    private final SubscriptionMetadata metadata;
+
+    private transient Subscription subscription;
+
+    private transient ServiceRegistration registration;
+
+    public CachedSubscription(SubscriptionMetadata metadata) {
+        notNull(metadata, "Subscription metadata cannot be null");
+        this.metadata = metadata;
+
+        this.subscription = null;
+        this.registration = null;
+    }
+
+    public SubscriptionMetadata getMetadata() {
+        return metadata;
+    }
+
+    public Optional<Subscription> getSubscription() {
+        return Optional.ofNullable(subscription);
+    }
+
+    public boolean isType(String type) {
+        return metadata.getTypeName()
+                .equals(type);
+    }
+
+    public boolean isNotType(String type) {
+        return !isType(type);
+    }
+
+    public synchronized boolean isNotRegistered() {
+        return registration == null;
+    }
+
+    /**
+     * Attempt to construct a {@link Subscription} using the provided factory. If the resultant
+     * subscription is valid, register it as this cached subscription's OSGi service.
+     *
+     * @param factory subscription factory corresponding to this cached subscription's type.
+     * @throws SubscriptionRegistrationException if the factory can't operate on this cached subscription.
+     */
+    public void registerSubscription(SubscriptionFactory factory) {
+        if (isNotType(factory.getTypeName())) {
+            throw new SubscriptionRegistrationException(format(
+                    "Factory type mismatch for subscription type [%s]",
+                    metadata.getTypeName()));
+        }
+
+        LOGGER.debug("Regenerating subscription [ {} | {} | {} ]",
+                metadata.getId(),
+                metadata.getTypeName(),
+                metadata.getCallbackAddress());
+
+        Subscription sub = factory.createSubscription(metadata);
+        if (sub == null) {
+            throw new SubscriptionRegistrationException("Unable to regenerate subscription");
+        }
+        registerSubscription(sub);
+    }
+
+    /**
+     * Register the provided {@link Subscription} as this cached subscription's OSGi service.
+     *
+     * @param sub the subscription to register.
+     * @throws SubscriptionRegistrationException if registration failed.
+     */
+    public synchronized void registerSubscription(Subscription sub) {
+        if (registration != null) {
+            throw new SubscriptionRegistrationException("Subscription already registered");
+        }
+
+        LOGGER.debug("Registering service [{}]", metadata.getId());
+
+        Dictionary<String, String> properties = new Hashtable<>(INITIAL_CAPACITY);
```

Comment:
Since this is API bound we should just mark this appropriate in SonarQube after this is merged. We can do so with the other findings as needed.